### PR TITLE
Connecting through a proxy

### LINKED ISSUE #427: Connect through a proxy
There are some use cases where users need to connect to the upstream server via a proxy that is not handling connections transparently.

I would like to add this functionality directly into Undici, and from the user perspective, it would look like this:
```js
const client = new Client({
  url: 'http://localhost:3000',
  proxy: 'http://localhost:3030'
})
```
Supporting this feature is also useful for working with proxy for debugging purposes, such as [mitmproxy](https://mitmproxy.org/).
I've implemented a similar solution for the Node.js's core HTTP client via a [custom agent](https://github.com/delvedor/hpagent).

What do you think?

### PR DESCRIPTION
**Pull request by delvedor**
This pr adds a test and the documentation to show how to connect through a proxy :)
I had to remove the super complex regex as it was not validating all absolute urls (eg: `http://localhost:8080` didn't pass the check). I think that lazily checking the protocol is enough, as trying to save *all users* from themselves is impossible.

/cc @szmarczak
Closes #427

### CODE REVIEW

**Review by ronag — APPROVED** [review]

File: docs/proxy.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+the `path` should be `path: 'http://upstream.server:port/hello?foo=bar'`.
+
+If you proxy requires basic authentication, you can send it via the `proxy-authorization` header.
+
+### Connect without authentication
+
+```js
+import { Client } from 'undici'
+import { createServer } from 'http'
+import proxy = from 'proxy'
```
Comment (mcollina) [inline docs/proxy.md:17]: this is not valid javascript but typescript.

**Review by mcollina — COMMENTED** [review]
Would you mind to show how to is with `Pool`?  

It would really be nice to also include some ways of having a `ProxyAgent`.

File: docs/proxy.md (hunk 1 above)
Comment (delvedor) [reply docs/proxy.md:17]: That's actually a typo! :D

**Comment by delvedor** [comment]
It works in the exact same way with both `Client` and `Pool`, so I've added an additional test and a note in the docs :)

**Comment by mcollina** [comment]
Here is an example for the Agent API:

```js
import { request, setGlobalAgent, ProxyAgent } from 'undici'

setGlobalAgent(new ProxyAgent('https://my.proxy.agent/'))

const {
  statusCode,
  headers,
  trailers,
  body
  // send the request via the Agent
} = await request('http://localhost:3000/foo')

console.log('response received', statusCode)
console.log('headers', headers)

for await (const data of body) {
  console.log('data', data)
}

console.log('trailers', trailers)
```

**Comment by mcollina** [comment]
FYI, you don't need to add the ProxyAgent right now. It can be another issue.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by budarin** [comment]
is there an example with proxying a POST request?

**Comment by mcollina** [comment]
No but it would be a good PR.

**Pull request merged**